[DashTree] Rework to improve live streaming

CastagnaIT opened the pull request:
## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here. -->
Its difficult to explain every single change here, i would end up writing a book!
So i try to summarize most important changes/problems

Unfortunately while i was fixing one thing it was breaking another, this due to wrong or missing implementations done in the past, since everything is connected there is no way to create separate small PRs you need to include all changes together

1) Nasty bug! on segment_buffers_ uses
https://github.com/xbmc/inputstream.adaptive/blob/df8b32db36cbf4e2e14441ebe533d547ebf3c947/src/common/AdaptiveStream.cpp#L882-L883
`segment_buffers_` is unrelated to MPD manifest updates! and can lead to wrong next segment!
because between MPD updates old segments can disappears from the timeline
and new one added, this will produce misaligned segment_number saved in to segment_buffers_
because the stored segment_number is not aware of MPD updates and keep the reference
to the initial (start number) segment position referred to oldest manifest!
with segment_buffers_ you need to find the next segment by compare segment startPTS.

2) Mess of segment timeline update on `CDashTree::RefreshLiveSegments`
https://github.com/xbmc/inputstream.adaptive/blob/df8b32db36cbf4e2e14441ebe533d547ebf3c947/src/parser/DASHTree.cpp#L1750-L1885
All this code was a lot messy, i have removed all, by looking one another video player source code, was checking to align updated segments by using segment start PTS, and not the start number as was done here. This should also allow more clean implementation for the misalignment workaround that before was force changing the start number, this was causing problems with some live streams and make working some others (like facebook), force change start number is a killer operation to do that may lead problems on many source code parts, often if you works with multiple periods

3) Nasty multiple bugs in the segment alignment code on representation segments (template) generation
https://github.com/xbmc/inputstream.adaptive/blob/df8b32db36cbf4e2e14441ebe533d547ebf3c947/src/parser/DASHTree.cpp#L1084-L1099
from the earliest versions of ISA i never understood this unclear code part, so it was used to adjust the start number for live streams, but...
after tons of try and check, i understood that:
3.1) Bug! this move the start number for the current live edge position, was ok, but if you move the number you must change also the start PTS! this last one was missing, and least to hidden weird behaviours
3.2) Bug! you can align the initial segment, but dont do it for all periods! this will kill the other period timeline! leading to unclear weird behaviours
3.3) Bug! you can align the initial segment, but if you dont know the period duration you will generate segments that can overlap up to and beyond the next period! with live streaming you need to wait the MPD update to know the period duration, so limit to generate the 1st segment and leaves the task to InsertLiveSegment
3.4) Bug! Dont align with MPD updates it its needed only for playback start or will affect the right periods timeline

4) Removed m_allowInsertLiveSegments workaround
the solution was if the SegmentTemplate provide the timeline dont insert live segments

5) InsertLiveSegment bug with multiple periods
"insert live segment's" when was used, was a non ending operation, that was preventing switch to the next period.
I found a way to limit the "live segment insert" that consist in to fetch the duration from MPD update during playback, and so i have add a code to check on-fly that inserted segment dont exceeds the period duration.
Also i tried to move the InsertLiveSegment call to the AdaptiveStream::ensureSegment instead of use Session,
this IMO will result in a more clean code

6) Reworked in full generation of segments for SegmentTemplate, there were wrong parts

7) `CSegment` `startPTS_` now is always referred to be within the period Start pts

8) ~Add support to MPD having SegmentTemplate and PresentationTimeOffset~ changes reverted/removed

9) Removed all Youtube outdated specific code about `$START_NUMBER$` on manifest update URL for MPD update and all code workarounds (`inputstream.adaptive.manifest_update_parameter` or `inputstream.adaptive.manifest_upd_params`)

10) `GetFragmentInfo` on `CSession::OnSegmentChangedRead` has been full removed, this because `InsertLiveSegment` is now done inside `AdaptiveStream::ensureSegment` that allow to remove specific code on Session like `m_hasSegmentChanged`

11) `AdaptiveStream::waitingForSegment` has been simplified and removed the weird `checkTime` var, i could not understand the reason for checking GetStreamType, where imo the only thing to take in account is make sure that available_segment_buffers_ is empty, if someone has some info please let me know, from my part i havent found regressions (i tested also HLS stream with audio "included" to video)

12) HLS add support to `EXT-X-PROGRAM-DATE-TIME` for live streams, this has been mandatory since now the `CSegment` `startPTS_` use PTS within the period, and so you need the datetime anchor (when possible)

13) some code TODO's added, for future cleanups

14) changes to gtest's. Some tests has been adjusted due to segment startPts fixes, two cases that may require some explanation are:
`DASHTreeTest updateParameterLiveSegmentStartNumber` removed was outdated youtube test
`DASHTreeAdaptiveStreamTest, MisalignedSegmentTimeline` i have reflected the changes to have the real segment timing as explained of old PR #724

15) Changes in to `CSubtitleSampleReader::ReadSample`, done because it was calling two times ensureSegment and was not so nice, so i have add a way to get the full data from segment buffer.

## Motivation and context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
I was investigating to fix the problem with `m_allowInsertLiveSegments` workaround
that since very old ISA versions was always handled in bad way,
i took time to try understand things that was not working good and i endedup finding a lot of bugs...

## How has this been tested?
<!--- Please describe in detail how you tested your change. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
This has been tested mainly with these live MPD,each one has different behaviour:
https://livesim.dashif.org/livesim/periods_20/testpic_2s/Manifest.mpd (+periods+insert segs)
https://d3cglye48cl6m2.cloudfront.net/out/v1/563ec3e9487340528616d736e1202232/index.mpd (+periods)
https://htv-rrs.mts.ru/PLTV/88888888/224/3221228154/3221228154.mpd (+insert segs, idr if there are periods maybe rare)
facebook live, mpd not linked find it on website flow (it expire) (NO insert segs, but with MPD updates old segments can disappears from timeline that can lead to  1° bug and others)

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the Wiki documentation
- [ ] I have updated the documentation accordingly

CastagnaIT commented:
hi @MoojMidge im partial rework dash support for live stream
and im stuck on a kind of exclusive ISA feature for Youtube DASH manifest
and so i need an your feedback to prevent broken some kind of your addon functionality

im talking of the use of `inputstream.adaptive.manifest_update_parameter` (or new `inputstream.adaptive.manifest_upd_params`) used with placeholder `$START_NUMBER$`
https://github.com/xbmc/inputstream.adaptive/wiki/Integration#inputstreamadaptivemanifest_update_parameter

its used to pass as argument the next segment number in the url to request live manifest update,
is this nowaday still used? or has it been removed from youtube backend?

In case it has been removed, i proceed to the complete removal of this exclusive youtube code
If not, I need you explain me way to test its operation since need to be reworked

MoojMidge commented:
@CastagnaIT - It can be removed. There are currently no specific url or query parameters required for the manifest updates to work. Normal full manifest updates work with a `$Number$` template now.

`<SegmentTemplate presentationTimeOffset="80326043" startNumber="19863" timescale="1000" media="sq/$Number$">`

CastagnaIT commented:
@MoojMidge thank you for feedback, so i can make code much more simpler

glennguy commented:
Wow looking forward to this.

DASH live refreshes have been totally broken for multi period streams since day 1 of multi period implementation. Doing the live HLS was hard enough, thinking about mpd gave me headaches!

CastagnaIT commented:
i need to investigate better about startPts/time set to generated segments
i realized that with a my recent code change i feel i have broken something

CastagnaIT commented:
@glennguy i have finally finished...
- i have reverted PTO uses since i have misinterpreted its use (should be used to sync streams at demuxer level)
- i have reverted all changes that have replaced `startPTS_` with `m_time`

i found additional bugs that affect pts on demuxers that will be not addressed here
i will add a comment to highlight it

CastagnaIT commented on src/samplereader/TSSampleReader.cpp:
```diff
@@ -69,6 +69,18 @@ AP4_Result CTSSampleReader::ReadSample()
 {
   if (ReadPacket())
   {
+    //! @todo: there is something wrong on pts calculation,
+    //! m_ptsOffs have a value in seconds and so the substraction "m_pts - m_ptsOffs" looks to be inconsistent,
+    //! To have pts in seconds on m_pts must be: pts = GetDts() / 90000,
+    //! but packet PTS seem to be different from m_ptsOffs pts value, as if it did not include the period start
+    //! so the substraction "m_pts - m_ptsOffs" it is not a clear thing.
+    //! There is also something weird on HLS discontinuities (multiple chapters/periods)
+    //! where after a discontinuity the packet pts is lower than the last segment of previous period/discontinuity
+    //! this cause a VP resync e.g:
+    //!   debug <general>: CVideoPlayer::CheckContinuity - resync backward :1, prev:587175999.000000, curr:577170666.000000, diff:-10005333.000000
+    //!   debug <general>: CVideoPlayer::CheckContinuity - update correction: -10026666.000000
+    //! not sure if this is correct or not (tested with pluto-tv)
+    //! This code is present also on the others sample readers, that need to be verified
     m_dts = (GetDts() == PTS_UNSET) ? STREAM_NOPTS_VALUE : (GetDts() * 100) / 9;
     m_pts = (GetPts() == PTS_UNSET) ? STREAM_NOPTS_VALUE : (GetPts() * 100) / 9;
```
pts/dts operations to me is full wrong, here and more likely also on all other demuxers/samplereaders
that result in a wrong pts/dts fed to kodi

glennguy replied:
Hmmm if the test was on pluto tv the segment lengths are 10 seconds, and the resync is 10 seconds (plus looks like difference between video/audio)... maybe old value is sent initially after discontinuity?

I still remember there is a bug (IMO) in VP when demuxread can't fill the internal 8s buffer fast enough on a period change, which causes VP to think it's out of sync by the same amount of time the entire stream has been playing. But that doesn't look like the case here. Could be related though? Something not being reinitialised to 0 in VP on period change and VP takes care of it this way usually?

CastagnaIT replied:
> Could be related though? Something not being reinitialised to 0 in VP on period change and VP takes care of it this way usually?

imo its different if you place a log in to "m_pts" value
on pluto you can see for example (random values)
packet pts 10000
packet pts 20000
packet pts 30000
---- discont - ads
packet pts 10000
packet pts 20000
---- discont - return to movie
packet pts ... a more higher value, now i dont remember exaclty if continue from last of 30000

idk if there is a way to intercept in right way these disc changes to prevent forward/backward of the VP that cause the resync
I have little info to understand if this is correct or if it needs proper fix

CastagnaIT commented on src/parser/DASHTree.cpp:
```diff
@@ -1021,127 +975,127 @@ void adaptive::CDashTree::ParseTagRepresentation(pugi::xml_node nodeRepr,
+            CSegment seg;
+            seg.startPTS_ = time;
+            // If no PTO, the "t" value on <SegmentTimeline><S> element should be relative to period start
+            if (!hasPTO)
+              seg.startPTS_ += periodStartScaled;
```
particular note to this `if (!hasPTO)` that add period start pts to segment pts,
this is to fix this sample stream:
```
#KODIPROP:inputstream=inputstream.adaptive
#KODIPROP:inputstream.adaptive.manifest_type=mpd
#KODIPROP:inputstream.adaptive.license_type=com.widevine.alpha
#KODIPROP:inputstream.adaptive.license_key=https://lic.staging.drmtoday.com/license-proxy-widevine/cenc/?specConform=true||R{SSM}|
https://d24rwxnt7vw9qb.cloudfront.net/v1/dash/e6d234965645b411ad572802b6c9d5a10799c9c1/All_Reference_Streams//6e16c26536564c2f9dbc5f725a820cff/index.mpd
```
usually segments have pts within the period, but this sample stream has some periods where segments pts are not within the period, I don't really understand why and I don't have enough info on how to deal with this.
maybe this sample stream is broken but i have no idea, atm i have no more sample stream to make a comparation so idk...

to me this sound more a workaround but i dont know, anyway its easy to remove, we'll see what happens

glennguy replied:
If by 'segment pts' you mean the pts of the sample data then this stream is probably trying to simulate having ads inserted which would usually never line up with PTS, compared to streams that just have 'chapters'. I don't know if this helps...

CastagnaIT replied:
> If by 'segment pts' you mean the pts of the

its from manifest, example:
```xml
<Period id="43047706_1" start="PT25610H40M24.076S">
       <AdaptationSet bitstreamSwitching="false" frameRate="25/1" mimeType="video/mp4" segmentAlignment="true" startWithSAP="1" subsegmentAlignment="true" subsegmentStartsWithSAP="1">
            <SegmentTemplate startNumber="1" timescale="90000"/>
            <Representation bandwidth="5000000" codecs="avc1.4d4028" height="1080" id="4" width="1920">
                <SegmentTemplate initialization="asset_1080_5_2init.mp4" media="asset_1080_5_2_$Number%09d$.mp4" startNumber="1" timescale="90000">
                    <SegmentTimeline>
                        <S d="180000" t="0"/>
```
you can see that "t" timestamp of the segment its "0" instead to be within the priod start
usually the "t" timestamp should be as follow:

```xml
<Period duration="PT86.600S" id="43047663" start="PT92198337.476S">
        <AdaptationSet bitstreamSwitching="true" id="792039359" mimeType="video/mp4" segmentAlignment="true" startWithSAP="1" subsegmentAlignment="true" subsegmentStartsWithSAP="1">
            <SegmentTemplate initialization="index_video_$RepresentationID$_0_init.mp4?m=1704732393" media="index_video_$RepresentationID$_0_$Number$.mp4?m=1704732393" presentationTimeOffset="4609916873831" startNumber="43047692" timescale="50000">
                <SegmentTimeline>
                    <S d="100000" r="13" t="4609919673831"/>
```
"t" == 4609919673831
4609919673831 / timescale (50000) = 92198393,476
period start is: 92198337,476
you can see that fall in the period

since in the example above the "t" is "0" segment fall in a bad timing
and so when you check for last segment with `AdaptiveTree::IsLastSegment` is not possible detect the last segment due to wrong segment pts, if you try to remove this code line, you can see that cause a playback problem

but this sample stream play bad also on web player testers, so idk if this is a sample stream bug or something i don't know/understand yet...

CastagnaIT replied:
i do the PR merge and release, we will see how to improve this when nevagive feedback will come

CastagnaIT committed:
[DashTree] Rework to improve live streaming

glennguy commented:
> @glennguy i have finally finished...
> 
>     * i have reverted PTO uses since i have misinterpreted its use (should be used to sync streams at demuxer level)
> 
>     * i have reverted all changes that have replaced `startPTS_` with `m_time`
> 
> 
> i found additional bugs that affect pts on demuxers that will be not addressed here i will add a comment to highlight it

Thanks, I reviewed last week so this saves a lot of time knowing what changed

glennguy reviewed (approved):
Awesome!

CastagnaIT committed:
[SubtitleSampleReader] Fix m_adStream nullptr access
